File: backend/app/schemas/ai_settings.py
"""ARCH-14 Step 8 CONTRACT: AI Settings Pydantic schemas without tenant cost fields."""

import uuid
from datetime import datetime
from enum import Enum
from typing import Union

from pydantic import BaseModel, Field, field_validator
from pydantic import ConfigDict


class AIProvider(str, Enum):
    GROQ = "GROQ"
    GEMINI = "GEMINI"


class AISettingsBase(BaseModel):
    """ARCH40-S1:ai-settings-ranges. The same contracts as the step 1 CHECKs.

    Before ARCH-40 nothing below the browser checked these. The database now
    does (ck_ai_settings_*_range), and this model states the same bounds so a
    bad value is refused as a 422 naming the field, not as a 500 from an
    IntegrityError. `verify_arch40.py` gate A7 asserts the two agree.
    """

    provider: AIProvider
    model: str = Field(min_length=1, max_length=100)
    temperature: float = Field(ge=0, le=2)
    max_output_tokens: int = Field(ge=1, le=32768)
    top_p: float = Field(ge=0, le=1)
    frequency_penalty: float = Field(ge=-2, le=2)
    presence_penalty: float = Field(ge=-2, le=2)

    @field_validator("model")
    @classmethod
    def _model_not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("model must not be blank")
        return value.strip()
    # ARCH40-S1:ai-settings-schema. The three dead fields are gone from the
    # wire, which is what stops the console re-introducing them: gate F1
    # asserts the form submits no field the backend schema lacks, checked
    # against the generated OpenAPI types.
    enable_streaming: bool


class AISettingsUpdate(AISettingsBase):
    pass


class AISettingsResponse(AISettingsBase):
    id: uuid.UUID
    workspace_id: uuid.UUID
    updated_by_user_id: Union[uuid.UUID, None] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/schemas/aiSettings.ts
import { z } from "zod";

/**
 * ARCH40-S2:ai-settings-schema. The same ranges the database now enforces
 * (ck_ai_settings_*_range) and the backend schema states. Before ARCH-40 this
 * was the only check anywhere, and the penalties were narrower here (0..2)
 * than the providers accept (-2..2). `verify_arch40.py` gate F1 compares
 * these keys with the backend's AISettingsBase.
 */
export const aiSettingsSchema = z.object({
    provider: z.enum([
        "GROQ",
        "GEMINI",
    ]),

    model: z
        .string()
        .trim()
        .min(1, "Model is required.")
        .max(100),

    temperature: z
        .number()
        .min(0)
        .max(2),

    max_output_tokens: z
        .number()
        .int()
        .min(1)
        .max(32768),

    top_p: z
        .number()
        .min(0)
        .max(1),

    frequency_penalty: z
        .number()
        .min(-2)
        .max(2),

    presence_penalty: z
        .number()
        .min(-2)
        .max(2),

    enable_streaming: z.boolean(),
});

export type AISettingsFormData = z.infer<
    typeof aiSettingsSchema
>;
